balancer-js/src/modules/liquidity-managment/migrations/builder.ts

```diff
@@ -0,0 +1,167 @@
+import { OutputReference, Relayer } from '@/modules/relayer/relayer.module';
+import * as actions from '@/modules/relayer/actions';
+import { buildPaths } from './helpers';
+import balancerRelayerAbi from '@/lib/abi/BalancerRelayer.json';
+import { Interface } from '@ethersproject/abi';
+import { BigNumber } from '@ethersproject/bignumber';
+
+const balancerRelayerInterface = new Interface(balancerRelayerAbi);
+
+/**
+ * Using array of objects to preserve the tokens order
+ */
+export interface MigrationPool {
+  address: string;
+  id?: string;
+  poolType?: string;
+  poolTypeVersion?: number;
+  tokens?: MigrationPool[];
+  mainIndex?: number;
+}
+
+export const migrationBuilder = (
+  user: string,
+  relayer: string,
+  bptAmount: string,
+  from: MigrationPool,
+  to: MigrationPool,
+  fromGauge?: string,
+  toGauge?: string
+): string => {
+  if (
+    !from.id ||
+    !to.id ||
+    !from.tokens ||
+    !to.tokens ||
+    !from.poolType ||
+    !to.poolType
+  ) {
+    throw 'Pool data is missing';
+  }
+
+  // Define tokens
+  const fromTokens = from.tokens.flatMap(({ address }) => address);
+  const toTokens = to.tokens.flatMap(({ address }) => address);
+
+  // Prefer proportional exit, except for ComposableStableV1
+  // Choose 0 as the exit token index
+  // TODO: make default exit token dynamic
+  const exitTokenIndex =
+    from.poolType === 'ComposableStable' && from.poolTypeVersion == 1 ? 0 : -1;
+
+  // Define output references
+  let exitOutputReferences: OutputReference[];
+  let swapOutputReferences: BigNumber[] = [];
+  if (exitTokenIndex > -1) {
+    exitOutputReferences = [
+      {
+        index: exitTokenIndex,
+        key: Relayer.toChainedReference(`10${exitTokenIndex}`),
+      },
+    ];
+    swapOutputReferences = [Relayer.toChainedReference(`20${exitTokenIndex}`)];
+  } else {
+    exitOutputReferences = fromTokens.map((_, idx) => ({
+      index: idx,
+      key: Relayer.toChainedReference(`10${idx}`),
+    }));
+    swapOutputReferences = fromTokens.map((_, idx) =>
+      Relayer.toChainedReference(`20${idx}`)
+    );
+  }
+
+  const joinAmount = Relayer.toChainedReference('999');
+
+  // Configure migration steps
+  const migrationSteps = [];
+  const needsExit = true;
+  const needsJoin = true;
+  let needsSwap = false;
+
+  if (from.poolType === 'ComposableStable') {
+    needsSwap = true;
```

Comment:
This would prevent ComposableStable > ComposableStable migrations for non-boosted pools right?

Reply:
These should work as well. Adding the swaps step will scan the tokens for these needing a swap. If there are none following join should just reuse exit amounts instead of swap amounts. I didn't test yet with pool compositions other than boosted, but it should eventually cover all of the cases.